src/mechanics/cell/CellIPData.h

```diff
@@ -13,17 +13,24 @@ template <int TDim>
 class CellIPData
 {
 public:
-    CellIPData(const DofContainer<ElementSimple*> elements, const NuTo::Jacobian<TDim>& jacobian,
+    CellIPData(const DofContainer<CellInterpolationBase*> cellInterpolation, const NuTo::Jacobian<TDim>& jacobian,
                const NaturalCoords& ipCoords)
-        : mElements(elements)
+        : mCellInterpolation(cellInterpolation)
         , mJacobian(jacobian)
         , mIPCoords(ipCoords)
     {
     }
 
     NMatrix GetNMatrix(const DofType& dofType) const
```

Comment:
IMO, there the N matrix is not possible to cache for IGA, since the shape functions are not only a function of the natural coordinate, but also from the knot vector. Certainly, this could be circumvented by making an IGA-cell and a FE cell. However, for a better generality of the code (and maybe sacrificing some performance) I would opt for just defining an interpolation that only provides shape functions and its derivatives. Everything else such as the calculation of B should be done somewhere else. I would even ask the question, if we really need to store these operators in a class such as cellIpData, or if this could not just be global operator function that gets the shape function derivatives and returns the B matrix (or any other operator). From my point of view, these operators are strongly related to the PDE, and not the cell interpolation, and should therefore not be stored at the interpolation routines.

Reply:
Sure, storing N for IGA is not feasible, not the question. But does a interpolation in IGA work with the N matrix? Say, for some boundary forces, is `Integrate(GetN() * pressure)` valid for IGA as well? Then it could be easily moved to `CellInterpolationBase`. `CellInterpolationIga` calculates N, `CellInterpolationFem` provides a precalculated N. Should be no problem, right?

And for me, this caching is crucial. And apart from that, I want to be able to define a `CellInterpolationFemRegularGrid` that is equal for all elements where everything is precalculated. Or maybe that is a special CellIpData class...

Reply:
Yes, the N matrix works in the same way as with standard FEM for the interpolation, thus calculating the equivalent force vector for a pressure function is also Integrate N * pressure. But isn't this already an specific application (integration of some mechanics force) that is more related to the PDE and not to the interpolation? But anyway, if there is  a significant performance boost, we can move this to interpolation base (and B as well). 
The definition of your own CellInterpolationFemRegularGrid with a fixed number of nodes, dimensions, integration and even a defined linear elastic law or even a hard coded stiffness should not be a problem at all. It would be interesting to see the performance differences.
That would by the way even work for the template concept, where the cell is derived from the integrand interface class, and then the operator includes an additional function that takes as an argument not the integrand but a cell. The only additional feature here is to distinguish between a cell interface and an ip interface.